### ISSUE TITLE
[feature] Conan 2.0 needs a package_type for applications with libraries

### ISSUE THREAD

**Issue by System-Arch** [conan-io/conan#12969]
### What is your suggestion?

Various applications such as GCC and Flex include headers and runtime libraries that are used by compiled or generated code. If one specifies the package_type as "application" for these package recipes, clients will not see their headers or libraries due to the following code in conans/model/requires.py:

>         if pkg_type is PackageType.APP:
>             # Change the default requires headers&libs to False for APPS
>             set_if_none("_headers", False)
>             set_if_none("_libs", False)
>             set_if_none("_run", True)

Thus Conan 2.0 would benefit from having an additional package_type "application-with-library" to support this use case. It could be implemented with the following one-line addition:

>     class PackageType(Enum):
>         LIBRARY = "library"  # abstract type, should contain shared option to define
>         STATIC = "static-library"
>         SHARED = "shared-library"
>         HEADER = "header-library"
>         BUILD_SCRIPTS = "build-scripts"
>         APP = "application"
>         APP_LIB = "application-with-library"    # Support applications with libraries (e.g., GCC, Flex, etc.)
>         PYTHON = "python-require"
>         UNKNOWN = "unknown"

I realize this is equivalent to not specifying any package_type but it is inconsistent to not be able to specify the package_type explicitly for every recipe. Doing so permits things like programmatically querying for all applications in a given code base, which becomes impossible when some of the applications are not labelled thusly. Thanks,

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @System-Arch 

Please check the review in https://github.com/conan-io/conan/pull/12970#pullrequestreview-1268871598, in short: adding new package types cannot be done without defining also the different default ``requirement traits`` and the different propagation rules for requirement traits from transitive dependencies.

**Comment by CD3** [comment]
I don't think this is necessary. If a project has an application and accompanying library, then you can just create two different packages, one for the application and one for the library. This is what other package managers like `apt` do.

On top of that, there would be the question of which context the application and library are supposed to be available in. Does the package contain an application that is going to generate some code that will then be compiled and linked against its library in the build context, or is the library supposed to be compiled and linked in the host context? Is the application supposed to run in the build or link context? Are the headers needed also, and in which context?

**Comment by hekrdla** [comment]
Hi I think that conan should somehow support `package_type` as component property and handle each component based on it.

One example can be `protobuf`, it contains both library(libprotobuf) and application(protoc). In this case, `libprotobuf` will be used in **host context** as library and `protoc` in **build context** as build tool.
